File: festival-web/src/main/java/cn/bdqfork/web/route/message/ContextParameterResolver.java
package cn.bdqfork.web.route.message;

import io.vertx.reactivex.core.MultiMap;
import io.vertx.reactivex.core.http.HttpServerRequest;
import io.vertx.reactivex.core.http.HttpServerResponse;
import io.vertx.reactivex.ext.web.RoutingContext;

import java.lang.reflect.Parameter;

/**
 * @author bdq
 * @since 2020/2/11
 */
public class ContextParameterResolver extends AbstractParameterResolver {
    @Override
    protected Object doResolve(Parameter parameter, RoutingContext routingContext) {
        Class<?> parameterType = parameter.getType();

        if (parameterType == RoutingContext.class) {
            return routingContext;
        }
        if (parameterType == HttpServerRequest.class) {
            return routingContext.request();
        }
        if (parameterType == HttpServerResponse.class) {
            return routingContext.response();
        }

        if (parameterType == MultiMap.class) {
            return resolveParams(routingContext);
        }

        return null;
    }

    @Override
    protected boolean resolvable(Parameter parameter) {
        Class<?> parameterType = parameter.getType();
        return parameterType == RoutingContext.class ||
                parameterType == HttpServerRequest.class ||
                parameterType == HttpServerResponse.class ||
                parameterType == MultiMap.class;

    }

}

File: festival-kotlin-web/src/main/kotlin/cn/bdqfork/kotlin/web/route/message/resolver/ContextParameterResolver.kt
package cn.bdqfork.kotlin.web.route.message.resolver

import io.vertx.core.MultiMap
import io.vertx.core.http.HttpServerRequest
import io.vertx.core.http.HttpServerResponse
import io.vertx.core.json.JsonObject
import io.vertx.ext.web.RoutingContext
import io.vertx.ext.web.Session
import java.lang.reflect.Parameter

/**
 * @author bdq
 * @since 2020/2/11
 */
class ContextParameterResolver : AbstractParameterResolver() {

    @Throws(Exception::class)
    override fun doResolve(parameter: Parameter, routingContext: RoutingContext): Any? {
        val parameterType = parameter.type
        if (parameterType == RoutingContext::class.java) {
            return routingContext
        }
        if (parameterType == HttpServerRequest::class.java) {
            return routingContext.request()
        }
        if (parameterType == HttpServerResponse::class.java) {
            return routingContext.response()
        }
        if (parameterType == Session::class.java) {
            return routingContext.session()
        }
        if (parameterType == MultiMap::class.java) {
            return resolveParams(routingContext)
        }
        if (parameterType == JsonObject::class.java) {
            val jsonObject = routingContext.bodyAsJson
            return jsonObject ?: JsonObject()
        }
        return null
    }

    override fun resolvable(parameter: Parameter, routingContext: RoutingContext): Boolean {
        val parameterType = parameter.type
        return parameterType == RoutingContext::class.java ||
                parameterType == HttpServerRequest::class.java ||
                parameterType == HttpServerResponse::class.java ||
                parameterType == MultiMap::class.java ||
                parameterType == JsonObject::class.java
    }

}